File: tree/binary_tree/0100_same_tree.js
/*
Given the roots of two binary trees p and q, write a function to check if they are the same or not.

Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.
*/

// https://leetcode.com/problems/same-tree/

var isSameTree = function (p, q) {
  // if both are null
  if (!p && !q) return true;
  // if one is null and the other is not
  if (!p || !q) return false;
  // if both are not null but have different values
  if (p.val !== q.val) return false;
  // check recursively
  return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
};

File: tree/traversal/0100_same_tree.py
# Given the roots of two binary trees p and q, write a function to check if they are the same or not.

# Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

# https://leetcode.com/problems/same-tree/

# list every situations
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        elif (p and not q) or (q and not p) or p.val != q.val:
            return False
        else:
            if not p.left and not p.right and not q.left and not q.right:
                return True
            elif (p.left and not q.left) or (p.right and not q.right):
                return False
            else:
                return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

# simplified: only list true situations          
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if p and q and p.val == q.val:
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
        # all false situations
        return False